File: server/Magav.Server/Services/AuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Security.Cryptography;
using System.Text;
using Magav.Common.Models.Auth;
using Magav.Server.Database;
using Microsoft.Extensions.Configuration;
using Microsoft.IdentityModel.Tokens;

namespace Magav.Server.Services;

public class AuthService
{
    private readonly MagavDbManager _dbManager;
    private readonly JwtSettings _jwtSettings;
    private readonly SecuritySettings _securitySettings;

    public AuthService(MagavDbManager dbManager, IConfiguration config)
    {
        _dbManager = dbManager ?? throw new ArgumentNullException(nameof(dbManager));
        _jwtSettings = config.GetSection("Jwt").Get<JwtSettings>()
            ?? throw new InvalidOperationException("JWT settings not configured");
        _securitySettings = config.GetSection("Security").Get<SecuritySettings>()
            ?? new SecuritySettings();
    }

    public async Task<LoginResponse> LoginAsync(string username, string password)
    {
        var user = await _dbManager.Users.GetByUserNameAsync(username);

        // Check if account is locked
        if (user?.LockoutUntil > DateTime.UtcNow)
            throw new AuthException("החשבון נעול. נסה שוב מאוחר יותר.");

        // Verify password using BCrypt (one-way hash comparison)
        if (user == null || !BCrypt.Net.BCrypt.Verify(password, user.PasswordHash))
        {
            if (user != null)
            {
                user.FailedLoginAttempts++;
                if (user.FailedLoginAttempts >= _securitySettings.MaxFailedLoginAttempts)
                    user.LockoutUntil = DateTime.UtcNow.AddMinutes(_securitySettings.LockoutMinutes);
                user.UpdatedAt = DateTime.UtcNow;
                await _dbManager.Users.UpdateAsync(user);
            }
            throw new AuthException("שם משתמש או סיסמה שגויים");
        }

        if (!user.IsActive)
            throw new AuthException("החשבון אינו פעיל");

        // Reset failed attempts on successful login
        user.FailedLoginAttempts = 0;
        user.LockoutUntil = null;
        user.LastConnected = DateTime.UtcNow;
        user.UpdatedAt = DateTime.UtcNow;

        // Generate tokens
        var accessToken = GenerateJwtToken(user);
        var refreshToken = GenerateSecureToken();
        var refreshTokenExpiry = DateTime.UtcNow.AddDays(_jwtSettings.RefreshTokenExpirationDays);

        // Store HASHED refresh token with expiry
        user.RefreshTokenHash = HashRefreshToken(refreshToken);
        user.RefreshTokenExpiry = refreshTokenExpiry;
        await _dbManager.Users.UpdateAsync(user);

        return new LoginResponse
        {
            AccessToken = accessToken,
            RefreshToken = refreshToken,
            ExpiresAt = DateTime.UtcNow.AddMinutes(_jwtSettings.AccessTokenExpirationMinutes).ToString("o"),
            User = new UserInfo
            {
                Id = user.Id.ToString(),
                Name = user.FullName,
                Email = $"{user.UserName}@magav.local",  // Placeholder email
                Roles = new[] { user.Role },
                Permissions = new Dictionary<string, object>()
            },
            MustChangePassword = user.MustChangePassword
        };
    }

    public async Task<LoginResponse> RefreshTokenAsync(string refreshToken)
    {
        var hash = HashRefreshToken(refreshToken);
        var user = await _dbManager.Users.GetByRefreshTokenHashAsync(hash);

        // Validate refresh token
        if (user == null || !user.IsActive)
            throw new AuthException("Invalid refresh token");

        // Check expiry
        if (user.RefreshTokenExpiry == null || user.RefreshTokenExpiry < DateTime.UtcNow)
            throw new AuthException("Refresh token expired");

        // Generate new tokens (rotate)
        var newAccessToken = GenerateJwtToken(user);
        var newRefreshToken = GenerateSecureToken();
        var newExpiry = DateTime.UtcNow.AddDays(_jwtSettings.RefreshTokenExpirationDays);

        user.RefreshTokenHash = HashRefreshToken(newRefreshToken);
        user.RefreshTokenExpiry = newExpiry;
        user.UpdatedAt = DateTime.UtcNow;
        await _dbManager.Users.UpdateAsync(user);

        return new LoginResponse
        {
            AccessToken = newAccessToken,
            RefreshToken = newRefreshToken,
            ExpiresAt = DateTime.UtcNow.AddMinutes(_jwtSettings.AccessTokenExpirationMinutes).ToString("o"),
            User = new UserInfo
            {
                Id = user.Id.ToString(),
                Name = user.FullName,
                Email = $"{user.UserName}@magav.local",
                Roles = new[] { user.Role },
                Permissions = new Dictionary<string, object>()
            },
            MustChangePassword = user.MustChangePassword
        };
    }

    public async Task LogoutAsync(int userId)
    {
        var user = await _dbManager.Users.GetByIdAsync(userId);
        if (user != null)
        {
            user.RefreshTokenHash = null;
            user.RefreshTokenExpiry = null;
            user.UpdatedAt = DateTime.UtcNow;
            await _dbManager.Users.UpdateAsync(user);
        }
    }

    // Generate JWT access token
    private string GenerateJwtToken(User user)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_jwtSettings.SecretKey));
        var credentials = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);

        var claims = new[]
        {
            new Claim(JwtRegisteredClaimNames.Sub, user.Id.ToString()),
            new Claim(JwtRegisteredClaimNames.Name, user.FullName),
            new Claim(ClaimTypes.Role, user.Role),
            new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
        };

        var token = new JwtSecurityToken(
            issuer: _jwtSettings.Issuer,
            audience: _jwtSettings.Audience,
            claims: claims,
            expires: DateTime.UtcNow.AddMinutes(_jwtSettings.AccessTokenExpirationMinutes),
            signingCredentials: credentials
        );

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    // Generate cryptographically secure random token
    private static string GenerateSecureToken()
    {
        var randomBytes = new byte[32];  // 256 bits
        using var rng = RandomNumberGenerator.Create();
        rng.GetBytes(randomBytes);
        return Convert.ToBase64String(randomBytes);
    }

    // Hash refresh token for secure storage
    private static string HashRefreshToken(string token) =>
        Convert.ToBase64String(SHA256.HashData(Encoding.UTF8.GetBytes(token)));
}

// Response models
public class ApiResponse<T>
{
    public bool Success { get; set; }
    public string? Message { get; set; }
    public T? Data { get; set; }
    public string[]? Errors { get; set; }

    public static ApiResponse<T> Ok(T data, string? message = null) => new()
    {
        Success = true,
        Data = data,
        Message = message
    };

    public static ApiResponse<T> Fail(string error) => new()
    {
        Success = false,
        Errors = new[] { error }
    };

    public static ApiResponse<T> Fail(string[] errors) => new()
    {
        Success = false,
        Errors = errors
    };
}

public class LoginResponse
{
    public string AccessToken { get; set; } = string.Empty;
    public string RefreshToken { get; set; } = string.Empty;
    public string ExpiresAt { get; set; } = string.Empty;  // ISO 8601
    public UserInfo User { get; set; } = new();
    public bool MustChangePassword { get; set; } = false;
}

public class UserInfo
{
    public string Id { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string Email { get; set; } = string.Empty;
    public string[] Roles { get; set; } = Array.Empty<string>();
    public Dictionary<string, object> Permissions { get; set; } = new();
}

// Settings classes
public class JwtSettings
{
    public string SecretKey { get; set; } = string.Empty;
    public string Issuer { get; set; } = string.Empty;
    public string Audience { get; set; } = string.Empty;
    public int AccessTokenExpirationMinutes { get; set; } = 15;
    public int RefreshTokenExpirationDays { get; set; } = 7;
}

public class SecuritySettings
{
    public int MaxFailedLoginAttempts { get; set; } = 5;
    public int LockoutMinutes { get; set; } = 15;
    public int BcryptWorkFactor { get; set; } = 12;
}

public class AuthException : Exception
{
    public AuthException(string message) : base(message) { }
}

File: android/app/src/main/java/com/magav/app/service/AuthService.kt
package com.magav.app.service

import at.favre.lib.crypto.bcrypt.BCrypt
import com.magav.app.api.auth.JwtConfig
import com.magav.app.api.models.LoginResponse
import com.magav.app.api.models.UserInfo
import com.magav.app.db.dao.UserDao
import java.security.MessageDigest
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

class AuthService(private val userDao: UserDao) {

    companion object {
        private const val MAX_FAILED_ATTEMPTS = 5
        private const val LOCKOUT_MINUTES = 15L
        private const val MIN_PASSWORD_LENGTH = 6
    }

    private val isoFormatter: DateTimeFormatter =
        DateTimeFormatter.ISO_INSTANT.withZone(ZoneOffset.UTC)

    suspend fun login(username: String, password: String): LoginResponse {
        val user = userDao.getByUserName(username)
            ?: throw IllegalArgumentException("שם משתמש או סיסמה שגויים")

        if (user.isActive != 1) {
            throw IllegalArgumentException("המשתמש אינו פעיל")
        }

        // Check lockout
        user.lockoutUntil?.let { lockoutStr ->
            val lockoutUntil = Instant.parse(lockoutStr)
            if (Instant.now().isBefore(lockoutUntil)) {
                throw IllegalArgumentException("החשבון נעול. נסה שוב מאוחר יותר")
            }
        }

        // Verify password
        val verified = BCrypt.verifyer()
            .verify(password.toCharArray(), user.passwordHash)
            .verified

        if (!verified) {
            val newAttempts = user.failedLoginAttempts + 1
            val now = isoFormatter.format(Instant.now())

            if (newAttempts >= MAX_FAILED_ATTEMPTS) {
                val lockoutUntil = isoFormatter.format(
                    Instant.now().plusSeconds(LOCKOUT_MINUTES * 60)
                )
                userDao.update(
                    user.copy(
                        failedLoginAttempts = newAttempts,
                        lockoutUntil = lockoutUntil,
                        updatedAt = now
                    )
                )
                throw IllegalArgumentException("החשבון ננעל עקב ניסיונות כושלים מרובים")
            } else {
                userDao.update(
                    user.copy(
                        failedLoginAttempts = newAttempts,
                        updatedAt = now
                    )
                )
                throw IllegalArgumentException("שם משתמש או סיסמה שגויים")
            }
        }

        // Successful login — reset attempts, update last connected
        val now = isoFormatter.format(Instant.now())
        val accessToken = JwtConfig.generateAccessToken(user.id, user.fullName, user.role)
        val refreshToken = JwtConfig.generateRefreshToken()
        val refreshTokenHash = hashSha256(refreshToken)
        val refreshTokenExpiry = isoFormatter.format(JwtConfig.getRefreshTokenExpiresAt().toInstant())

        userDao.update(
            user.copy(
                failedLoginAttempts = 0,
                lockoutUntil = null,
                lastConnected = now,
                refreshTokenHash = refreshTokenHash,
                refreshTokenExpiry = refreshTokenExpiry,
                updatedAt = now
            )
        )

        return LoginResponse(
            accessToken = accessToken,
            refreshToken = refreshToken,
            expiresAt = JwtConfig.getExpiresAt(),
            user = UserInfo(
                id = user.id.toString(),
                name = user.fullName,
                roles = listOf(user.role)
            ),
            mustChangePassword = user.mustChangePassword == 1
        )
    }

    suspend fun refreshToken(refreshToken: String): LoginResponse {
        val hash = hashSha256(refreshToken)
        val user = userDao.getByRefreshTokenHash(hash)
            ?: throw IllegalArgumentException("טוקן לא תקף")

        // Check expiry
        user.refreshTokenExpiry?.let { expiryStr ->
            val expiry = Instant.parse(expiryStr)
            if (Instant.now().isAfter(expiry)) {
                // Clear expired token
                val now = isoFormatter.format(Instant.now())
                userDao.update(
                    user.copy(
                        refreshTokenHash = null,
                        refreshTokenExpiry = null,
                        updatedAt = now
                    )
                )
                throw IllegalArgumentException("טוקן פג תוקף")
            }
        } ?: throw IllegalArgumentException("טוקן לא תקף")

        // Rotate tokens
        val now = isoFormatter.format(Instant.now())
        val newAccessToken = JwtConfig.generateAccessToken(user.id, user.fullName, user.role)
        val newRefreshToken = JwtConfig.generateRefreshToken()
        val newRefreshTokenHash = hashSha256(newRefreshToken)
        val newRefreshTokenExpiry = isoFormatter.format(JwtConfig.getRefreshTokenExpiresAt().toInstant())

        userDao.update(
            user.copy(
                refreshTokenHash = newRefreshTokenHash,
                refreshTokenExpiry = newRefreshTokenExpiry,
                lastConnected = now,
                updatedAt = now
            )
        )

        return LoginResponse(
            accessToken = newAccessToken,
            refreshToken = newRefreshToken,
            expiresAt = JwtConfig.getExpiresAt(),
            user = UserInfo(
                id = user.id.toString(),
                name = user.fullName,
                roles = listOf(user.role)
            ),
            mustChangePassword = user.mustChangePassword == 1
        )
    }

    suspend fun logout(userId: Int) {
        val user = userDao.getById(userId) ?: return
        val now = isoFormatter.format(Instant.now())
        userDao.update(
            user.copy(
                refreshTokenHash = null,
                refreshTokenExpiry = null,
                updatedAt = now
            )
        )
    }

    suspend fun changePassword(userId: Int, newPassword: String) {
        if (newPassword.length < 4) {
            throw IllegalArgumentException("הסיסמה חייבת להכיל לפחות 4 תווים")
        }

        val user = userDao.getById(userId)
            ?: throw IllegalArgumentException("משתמש לא נמצא")

        val now = isoFormatter.format(Instant.now())
        val newHash = BCrypt.withDefaults().hashToString(10, newPassword.toCharArray())

        userDao.update(
            user.copy(
                passwordHash = newHash,
                mustChangePassword = 0,
                updatedAt = now
            )
        )
    }

    private fun hashSha256(input: String): String {
        return MessageDigest.getInstance("SHA-256")
            .digest(input.toByteArray())
            .joinToString("") { "%02x".format(it) }
    }
}
